parlai/tasks/flattened/agents.py

```diff
@@ -0,0 +1,232 @@
+        ex.force_set('episode_done', True)
+        labels = ex.get('labels', ex.get('eval_labels', None))
+        if labels is not None and include_labels:
+            context.append(random.choice(labels))
+
+        new_episode.append(ex)
+
+    return new_episode
+
+
+def get_original_task(opt, multi_possible=False):
+    modules = []
+    tasks = opt['task'].split(',')
+    if not multi_possible:
+        assert len(tasks) == 1
+
+    for task in tasks:
+        if len(task.split(':')) < 3:
+            raise RuntimeError(
+                '\n\n********************************************************\n'
+                'Must specify original task using the following format:\n'
+                '`--task flattened:task:<ORIGINAL TASK NAME>`'
+                '\n********************************************************\n'
+            )
+        original_task = ':'.join(task.split(':')[2:])
+        task_module = get_task_module(original_task)
+        modules.append(task_module)
+
+    if multi_possible:
+        return modules
+
+    return modules[0]
+
+
+class TaskTeacher(FixedDialogTeacher):
+    """
+    Predict the gender of character given the dialogue utterance.
+    """
+
+    @staticmethod
+    def add_cmdline_args(parser):
+        agent = parser.add_argument_group('Flattened Teacher Args')
+        agent.add_argument(
+            '--flatten-include-labels',
+            type='bool',
+            default=True,
+            help='Include labels in the history when flattening an episode',
+        )
+        agent.add_argument(
+            '--flatten-delimiter',
+            type=str,
+            default='\n',
+            help='How to join the dialogue history from previous turns.',
+        )
+        agent.add_argument(
+            '--max-context-length', type=int, default=-1, help='Maximum context length'
+        )
+        agent.add_argument(
+            '--invalidate-cache',
+            type='bool',
+            default=False,
+            help='Set this to True to rebuild the data',
+        )
+
+        # Add the arguments for the teacher
+        opt = parser.parse_and_process_known_args()[0]
+        tasks = get_original_task(opt, multi_possible=True)
+        for task in tasks:
+            if hasattr(task, 'add_cmdline_args'):
+                task.add_cmdline_args(parser)
+
+        return parser
+
+    def __init__(self, opt, shared=None):
+        self.opt = opt
+
+        if shared and 'data' in shared:
+            self.data = shared['data']
+        else:
+            self.data = self._setup_data(opt)
+
+        super().__init__(opt, shared)
+        self.reset()
+
+    def num_episodes(self):
+        return len(self.data)
+
+    def num_examples(self):
+        return len(self.data)
+
+    def _setup_data(self, opt, context_length=-1):
+        # possibly make new data directory
+        original_task = get_original_task(opt)
+        self.original_task = ':'.join(opt['task'].split(':')[2:])
```

Comment:
could we return the `original_task` from `get_original_task` to not have to redo this computation (and possibly rename that function to indicate it's returning the task module)